web/server/network_check.go

```diff
@@ -0,0 +1,326 @@
+	logger := logging.NewLogger("network-checks")
+	if testing.Testing() {
+		logger.Debug("Skipping network checks in a testing environment")
+		return
+	}
+
+	online := testGeneral(ctx)
+	if !online {
+		logger.Warn("Machine appears to be offline (cannot make connection to app.viam.com); skipping further network checks")
+		return
+	}
+
+	if err := testUDP(ctx, logger.Sublogger("udp")); err != nil {
+		logger.Errorw("Error running general network tests", "error", err)
+	}
+
+	if err := testTCP(ctx, logger.Sublogger("tcp")); err != nil {
+		logger.Errorw("Error running general network tests", "error", err)
+	}
+}
+
+// Tests general network connectivity (to app.viam.com.) Stolen from datamanager's
+// sync/connectivity.go code. Returns false is offline and true if online.
+func testGeneral(ctx context.Context) bool {
+	timeout := 5 * time.Second
+	attempts := 1
+	if runtime.GOOS == "windows" {
+		// TODO(RSDK-8344): this is temporary as we 1) debug connectivity issues on windows,
+		// and 2) migrate to using the native checks on the underlying connection.
+		timeout = 15 * time.Second
+		attempts = 2
+	}
+
+	for i := range attempts {
+		// Use DialDirectGRPC to make a connection to app.viam.com instead of a
+		// basic net.Dial in order to ensure that the connection can be made
+		// behind wifi or the BLE-SOCKS bridge (DialDirectGRPC can dial through
+		// the BLE-SOCKS bridge.)
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		conn, err := rpc.DialDirectGRPC(ctx, "app.viam.com:443", nil)
+		cancel()
+		if err == nil {
+			conn.Close() //nolint:gosec,errcheck
+			return true
+		}
+		if i < attempts-1 {
+			time.Sleep(time.Second)
+		}
+	}
+
+	return false
+}
+
+var (
+	stunServerURLsToTestUDP = []string{
+		"global.stun.twilio.com:3478",
+		"turn.viam.com:443",
+		"stun.l.google.com:3478",
+		"stun.l.google.com:19302",
+		"stun.sipgate.net:3478",
+		"stun.sipgate.net:3479",
+	}
+	stunServerURLsToTestTCP = []string{
+		"turn.viam.com:443", // only STUN server that acceps TCP STUN traffic
+	}
+)
+
+// Tests NAT over UDP against STUN servers.
+func testUDP(ctx context.Context, logger logging.Logger) error {
+	// Listen on arbitrary UDP port.
+	conn, err := net.ListenPacket("udp", "0.0.0.0:0")
+	if err != nil {
+		logger.Warn("Failed to listen over UDP on a port; UDP traffic may be blocked")
+		return err
+	}
+
+	// `net.PacketConn`s do not function with contexts (only deadlines.) If passed-in
+	// context expires (machine is likely shutting down,) _or_ tests finish, close the
+	// underlying `net.PacketConn` asynchronously to stop ongoing network checks.
+	testUDPDone := make(chan struct{})
+	defer func() {
+		testUDPDone <- struct{}{}
+	}()
+	go func() {
+		select {
+		case <-ctx.Done():
+		case <-testUDPDone:
+		}
+		conn.Close() //nolint:gosec,errcheck
+	}()
+
+	// Set a deadline for all interactions of 10 seconds in the future.
+	if err := conn.SetDeadline(time.Now().Add(10 * time.Second)); err != nil {
```

Comment:
This needs to be a `SetReadDeadline` and it needs to be set on each iteration -- not for the whole test. Being unable to talk with one stun server should not abort running the remaining tests. That information is critical for diagnosis.